[1.5 Release][Dist Autograd][Better Engineering] Notify Workers on Failure during Distributed Autograd

osalpekar opened the pull request:
Summary:
Fixes: https://github.com/pytorch/pytorch/issues/27643

This PR manages notifying workers in the event of a failure during distributed autograd. Gracefully handles propagating errors across all nodes in the backward pass and sets state in the local autograd engines accordingly.

Test Plan: Added 2 new tests checking errors when they are thrown in an intermediate node during distributed autograd. Ensured that all existing distributed autograd tests pass.

Differential Revision: D20164420

pritamdamania87 commented on torch/csrc/distributed/autograd/context/context.h:
```diff
@@ -82,6 +87,12 @@ class TORCH_API DistAutogradContext {
+  // neighbors of errors during autograd is recursive, it is possible that this
+  // function is called multiple times. Thus, it should not crash if the
+  // GraphTask has already been set with an error by a previous RPC.
+  std::shared_ptr<torch::autograd::GraphTask> retrieveGraphTaskIdempotent();
```
nit: call this `retrieveGraphTaskIfExists`

pritamdamania87 commented on torch/csrc/distributed/autograd/context/container.h:
```diff
@@ -82,6 +82,12 @@ class TORCH_API DistAutogradContainer {
 
+  // Retrieves the autograd context, if it exists, and sets the autograd error
+  // on that context when there is an autograd failure.
+  void setAutogradContextErrorIfPresent(
```
Doesn't seem right to add this method to the container. Can we do this in request_callback_impl.cpp instead?

1. Retrieve the context from the container.
2. If it exists, retrieve the graph task. If the graph task is valid set an exception on it.
3. If the context doesn't exist or graph task is not valid, just propagate the error to all other nodes.

pritamdamania87 commented on torch/csrc/distributed/autograd/context/context.cpp:
```diff
@@ -193,6 +202,28 @@ std::shared_ptr<rpc::FutureMessage> DistAutogradContext::
+void DistAutogradContext::setAutogradError(const std::string& errorMsg) {
+  setGraphTaskException(errorMsg);
+  propagateAutogradError(errorMsg);
+}
```
should be done in request_callback_impl.cpp by just retrieving information from the context.

pritamdamania87 commented on torch/csrc/distributed/autograd/context/context.cpp:
```diff
@@ -193,6 +202,28 @@ std::shared_ptr<rpc::FutureMessage> DistAutogradContext::
+  if (graphTask) {
+    std::runtime_error err(errorMsg);
+    graphTask->set_exception(err, nullptr);
+  }
```
This should move to request_callback_impl.cpp. Also, note that you are not holding the lock here.

osalpekar replied:
I kept this helper function here for now (updated to hold the lock like we do in `addOutstandingRpc` when setting an exception on `graphTask`). Should we still move it to `request_callback_impl.cpp` and set the exception outside of the lock?

pritamdamania87 commented on torch/csrc/distributed/autograd/rpc_messages/autograd_backward_failure_req.cpp:
```diff
@@ -0,0 +1,64 @@
+      payload_size,
+      *rpc::RpcAgent::getCurrentRpcAgent()->getTypeResolver(),
+      &message.tensors());
+  std::vector<at::IValue> tupleElements = tuple.toTuple()->elements();
```
nit: add TORCH_INTERNAL_ASSERT that we have exactly two elements in the tuple.

pritamdamania87 commented on torch/csrc/distributed/autograd/rpc_messages/autograd_backward_failure_req.h:
```diff
@@ -0,0 +1,30 @@
+namespace autograd {
+
+// Used to notify other workers when there is an autograd error.
+class TORCH_API AutogradBackwardFailureReq : public rpc::RpcCommandBase {
```
nit: call this `DistautogradFailureReq` since AutogradBackward might be confused with the local autograd engine.

pritamdamania87 commented on torch/csrc/distributed/rpc/message.h:
```diff
@@ -44,6 +44,9 @@ enum MessageType {
   CLEANUP_AUTOGRAD_CONTEXT_REQ = 19,
   CLEANUP_AUTOGRAD_CONTEXT_RESP = 20,
 
+  AUTOGRAD_BACKWARD_FAILURE_REQ = 21,
```
nit: `DIST_AUTOGRAD_FAILURE_REQ`.

pritamdamania87 commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -302,11 +304,21 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
-              autogradMetadata.autogradContextId);
+      std::shared_ptr<DistAutogradContext> autogradContext;
+
+      // In rare cases, a Backward_autograd request may arrive after the
```
nit: `BACKWARD_AUTOGRAD`

pritamdamania87 commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -302,11 +304,21 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
+      // gradient computations should take place on this context.
+      try {
+        autogradContext = DistAutogradContainer::getInstance().retrieveContext(
+            autogradMetadata.autogradContextId);
```
Add retrieveContextIfExists to the container instead. Its not clean catching the exception here since someone could modify retrieveContext to throw an exception for a different reason and now you're catching that exception assuming its because the context is not valid.

pritamdamania87 commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -302,11 +304,21 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
+        LOG(INFO) << "Ignoring Backward Autograd Request. Context "
+                  << autogradMetadata.autogradContextId
+                  << " already cleaned up due to autograd error";
+        return std::make_shared<rpc::FutureMessage>();
```
Don't we need to mark this Future as completed? Otherwise this future will never be marked as complete and the RPC layer will never respond back to the original request.

pritamdamania87 commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -302,11 +304,21 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
+        autogradContext = DistAutogradContainer::getInstance().retrieveContext(
+            autogradMetadata.autogradContextId);
+      } catch (std::exception& e) {
+        LOG(INFO) << "Ignoring Backward Autograd Request. Context "
```
If possible, would be nice to have a unit test covering this case.

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -1169,6 +1169,63 @@ def test_backward_autograd_engine_error(self):
+                RuntimeError, "Simulate error on backward pass"
+            ):
+                # Run backwards, and validate we receive an error.
+                dist_autograd.backward(context_id, [val.sum()])
```
We should validate that all context ids were cleaned up on all nodes after this.

pritamdamania87 commented on torch/csrc/distributed/autograd/engine/dist_engine.h:
```diff
@@ -7,6 +7,7 @@
 #include <torch/csrc/autograd/function.h>
 #include <torch/csrc/autograd/functions/basic_ops.h>
 #include <torch/csrc/distributed/autograd/context/context.h>
+#include <torch/csrc/distributed/autograd/rpc_messages/autograd_backward_failure_req.h>
```
Do we need this include here?

pritamdamania87 commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -210,6 +210,7 @@ std::shared_ptr<rpc::FutureMessage> DistEngine::runEngineAndAccumulateGradients(
               DistAutogradContainer::getInstance().getWorkerId(),
               ": ",
               error->what());
+          autogradContext->setAutogradError(errorMsg);
```
Shouldn't we just call `propagateAutogradError` since this node has already received an exception, so there is no point setting an exception on the GraphTask again.

Also, isn't it sufficient to do this just in DistEngine::Execute? Since the node that drives distributed autograd, always receives an error and that node can propagate the error. In DistEngine::Execute() instead of just waiting on the future, you can add a callback to the future which propagates the error and then wait on the future.

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -1169,6 +1169,63 @@ def test_backward_autograd_engine_error(self):
                 # Run backwards, and validate we receive an error.
                 dist_autograd.backward(context_id, [val.sum()])
 
+    @dist_init
```
Can we remove this hack since we have the proper fix now: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/distributed/rpc/dist_autograd_test.py#L1558? I guess removing rpc.shutdown() and sys.exit() from that test should work. Would be nice to run that test 100 times just to make sure its not flaky after those changes.

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -1169,6 +1169,63 @@ def test_backward_autograd_engine_error(self):
+                "worker{}".format(self._next_rank()), torch.div, args=(val, t2)
+            )
+
+            dist_autograd.backward(context_id, [val.sum()])
```
What exactly are we testing here? We have multiple tests in this file which test the backward without any errors. Do we still need this test?

pritamdamania87 reviewed (changes requested):
Awesome work!

pritamdamania87 commented on torch/csrc/distributed/autograd/context/context.cpp:
```diff
@@ -193,6 +202,25 @@ std::shared_ptr<rpc::FutureMessage> DistAutogradContext::
   return state->future;
 }
 
+void DistAutogradContext::setGraphTaskException(const std::string& errorMsg) {
```
nit: can we use this function in `addOutstandingRpc` in the futureMessage callback?

pritamdamania87 commented on torch/csrc/distributed/autograd/context/context.cpp:
```diff
@@ -193,6 +202,25 @@ std::shared_ptr<rpc::FutureMessage> DistAutogradContext::
 }
 
+void DistAutogradContext::setGraphTaskException(const std::string& errorMsg) {
+  std::runtime_error err(errorMsg);
```
Why are we doing this? Looks like we convert errorMsg to runtim_error then just get the error message back in err.what() ?

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -1169,6 +1169,39 @@ def test_backward_autograd_engine_error(self):
+            ):
+                # Run backwards, and validate we receive an error.
+                dist_autograd.backward(context_id, [val.sum()])
+
```
Can we add the number of `DIST_AUTOGRAD_FAILURE_REQ` messages received per context id to debug info and then verify here that each node gets one of these messages for all the other context ids? Can add this as a separate PR.

osalpekar committed:
[1.5 Release][Dist Autograd][Better Engineering] Notify Workers on Failure during Distributed Autograd (#34638)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/34638

Fixes: https://github.com/pytorch/pytorch/issues/27643

This PR manages notifying workers in the event of a failure during distributed autograd. Gracefully handles propagating errors across all nodes in the backward pass and sets state in the local autograd engines accordingly.

Test Plan: Added 2 new tests checking errors when they are thrown in an intermediate node during distributed autograd. Ensured that all existing distributed autograd tests pass.

Differential Revision: D20164420

fbshipit-source-id: 5aada5544ed12cd7e24053ba3b93f8b9b38ba021

ezyang commented:
this broke lint

```
  {
    path: 'test/test_jit.py',
    start_line: 8424,
    end_line: 8424,
    start_column: 33,
    end_column: 33,
    annotation_level: 'failure',
    message: '[E999] SyntaxError: invalid syntax'
  },
  {
    path: 'torch/testing/_internal/distributed/rpc/dist_autograd_test.py',
    start_line: 1,
    end_line: 1,
    start_column: 1,
    end_column: 1,
    annotation_level: 'failure',
    message: "[F401] 'sys' imported but unused"
  }
```
